tests/Agency/StorePerformanceTest.cpp

```diff
@@ -0,0 +1,423 @@
+      ss << ",[{\"k" << i << "\": " << i << "}]";
+    }
+    ss << "]";
+    writeAndCheck(ss.str());
+  }
+  std::stringstream ss;
+  ss << "[[";
+  ss << "\"k0\"";
+  for (int i{1}; i < hammer_times; ++i) {
+    ss << ",\"k" << i << "\"";
+  }
+  ss << "]]";
+
+  auto const key_list {ss.str()};
+
+  for (int i{}; i < hammer_times; ++i) {
+    auto const result {read(key_list)};
+    ASSERT_EQ(result->size(), result->size());
+  }
+}
+
+TEST_F(StorePerformanceTest, bigger_tx_w) {
+  std::stringstream ss;
+  ss << "[[{\"k0\": 0}]";
+  for (int i{1}; i < hammer_times; ++i) {
+    ss << ",[{\"k" << i << "\": " << i << "}]";
+  }
+  ss << "]";
+  auto const write_string {ss.str()};
+  for (int i{}; i < hammer_times; ++i) {
+    write(write_string);
+  }
+}
+
+// test array operations specifically
+
+TEST_F(StorePerformanceTest, array_push) {
+  writeAndCheck(R"([[{"/a/b/c":[]}]])");
+  for (int i{}; i < hammer_times; ++i) {
+    writeAndCheck(R"([[{"/a/b/c":{"op":"push","new":)" + std::to_string(i) + "}}]]");
+  }
+}
+
+TEST_F(StorePerformanceTest, array_pop) {
+  std::stringstream ss;
+  ss << R"([[{"/a/b/c":[0)";
+  for (int i{1}; i < hammer_times; ++i) {
+    ss << "," << i;
+  }
+  ss << "]}]]";
+  writeAndCheck(ss.str());
+  for (int i{}; i < hammer_times; ++i) {
+    writeAndCheck(R"([[{"/a/b/c":{"op":"pop"}}]])");
+  }
+}
+
+// test object operations specifically
+
+// test operations which need to change a lot in the tree
+// add in ascending order
+TEST_F(StorePerformanceTest, tree_add_ascending) {
+  auto const width {log10(hammer_times)};
+  for (int i{}; i < hammer_times; ++i) {
+    auto k {std::to_string(i)};
+    k = std::string(width - k.size(), '0') + k;
+    write("[[{\"k"+k+"\": 42}]]");
+  }
+}
+
+// add in descending order
+TEST_F(StorePerformanceTest, tree_add_descending) {
+  auto const width {log10(hammer_times)};
+  for (int i{hammer_times}; i > 0; --i) {
+    auto k {std::to_string(i)};
+    k = std::string(width - k.size(), '0') + k;
+    write("[[{\"k"+k+"\": 42}]]");
+  }
+}
+
+// add ascending, remove root, then re-add
+TEST_F(StorePerformanceTest, tree_add_remove_readd) {
+  auto const width {log10(hammer_times)};
+  for (int i{}; i < hammer_times; ++i) {
+    auto k {std::to_string(i)};
+    k = std::string(width - k.size(), '0') + k;
+    write("[[{\"/t/k"+k+"\": 42}]]");
+  }
+  write("[[{\"/t\": 0}]]");
+}
+
+// test for contention:
+// multiple threads manipulate values
+TEST_F(StorePerformanceTest, multiple_threads_all_separate_keys) {
+  std::vector<std::thread> threads;
+  std::generate_n(std::back_inserter(threads), hammer_times, [this](){return std::thread([this](){
+    auto const key {rand_path()};
+    auto const json_object {"{\"" + key + "\": " + std::to_string(rand()) + "}"};
+    auto result {write(std::vector<std::vector<std::string>>{{json_object}})};
+    ASSERT_EQ(result.front(), consensus::apply_ret_t::APPLIED);
+  });});
+  for (auto& thread: threads) {
```

Comment:
I think we should make sure that all threads start at the same time and use a condition variable or barrier to let them actually start work at the same time.

Reply:
I would be interested in individual time measurements of each thread in addition to an overall measurement.

Reply:
It is a mistake to do only one call to the Store in each thread. In this way, thread creation and joining will vastly dominate the actual execution time rendering any measurement useless to judge the actual Store performance. I think we should vary the number of threads and each thread should - concurrently - run many executions. We probably also need some kind of multi-threaded protection, since the Store itself is not thread-safe as far as I remember.

Reply:
This last point is truly important; will address it.